src/runtime_src/core/pcie/linux/plugin/xdp/hal_profile.cpp

```diff
@@ -386,36 +402,33 @@ void load_xdp_plugin_library(HalPluginConfig* )
     if(xrt_core::config::get_profile()) {
       // "profile=true" is also set. This enables OpenCL based flow for profiling. 
       // Currently, mix of OpenCL and HAL based profiling is not supported.
-      xrt_core::message::send(xrt_core::message::severity_level::XRT_WARNING, "XRT", std::string("Both profile=true and xrt_profile=true set in xrt.ini config. Currently, these flows are not supported to work together."));
+      xrt_core::message::send(xrt_core::message::severity_level::XRT_WARNING, "XRT",
+                std::string("Both profile=true and xrt_profile=true set in xrt.ini config. Currently, these flows are not supported to work together."));
       return;
     }
 
     bfs::path xrt(emptyOrValue(getenv("XILINX_XRT")));
-    bfs::path libname("libxdp_hal_plugin.so");
     if (xrt.empty()) {
-      xrt_core::message::send(xrt_core::message::severity_level::XRT_ERROR, "XRT", std::string("Library " + libname.string() + " not found! XILINX_XRT not set"));
-      exit(EXIT_FAILURE);
-    }
-    bfs::path p(xrt / "lib");
-    if(directoryOrError(p)) {
-      exit(EXIT_FAILURE);
+      throw std::runtime_error("Library xdp_hal_plugin not found! XILINX_XRT not set");
     }
-    p /= libname;
-    if (!isDLL(p)) {
-      xrt_core::message::send(xrt_core::message::severity_level::XRT_ERROR, "XRT", std::string("Library " + p.string() + " not found!"));
-      exit(EXIT_FAILURE);
-    }
-    auto handle = dlopen(p.string().c_str(), RTLD_NOW | RTLD_GLOBAL);
-    if (!handle) {
-      xrt_core::message::send(xrt_core::message::severity_level::XRT_ERROR, "XRT", std::string("Failed to open XDP hal plugin library '" + p.string() + "'\n" + dlerror()));
-      exit(EXIT_FAILURE);
+    bfs::path xrtlib(xrt / "lib");
+    directoryOrError(xrtlib);
+    auto libname = dllpath(xrt, "xdp_hal_plugin");
+    if (!isDLL(libname)) {
+      throw std::runtime_error("Library " + libname.string() + " not found!");
     }
+    auto handle = xrt_core::dlopen(libname.string().c_str(), RTLD_NOW | RTLD_GLOBAL);
+    if (!handle)
+      throw std::runtime_error("Failed to open XDP HAL Profile library '" + libname.string() + "'\n" + xrt_core::dlerror());
+
     const std::string cb_func_name = "hal_level_xdp_cb_func";
-    dlerror();
-    cb = cb_func_type(reinterpret_cast<cb_load_func_type>(dlsym(handle, cb_func_name.c_str())));
-    if(dlerror() != NULL) { // check if dlsym was successful
+    xrt_core::dlerror();
+    cb = cb_func_type(reinterpret_cast<cb_load_func_type>(xrt_core::dlsym(handle, cb_func_name.c_str())));
+#ifndef _WIN32
+    if(xrt_core::dlerror() != NULL) { // check if dlsym was successful
       cb = nullptr;
     }
+#endif
```

Comment:
shouldn't you just check 'if (!cb)' both windows and linux?

Reply:
I saw that for Windows, xrt_core::dlerror returns "". So, on Windows (xrt_core::dlerror() !=NULL) will be true resulting in cb = nullptr, even if the xrt_core::dlsym succeeded.
Shall I add change in xrt_core::dlerror ?

Reply:
No, my point was that you should check the return value of xrt_core:dlsym; it it is nullptr, then it failed, same on both linux and and windows.  If the dlsym returns nullptr, then check the error by calling dlerror.

Reply:
okay. I have updated.